### PR TITLE
feat: Add support for ordered "Steps"

### PR DESCRIPTION
**Pull request by grahammcculloch**
Connects-to: #1077
Change-type: minor
Signed-off-by: Graham McCulloch <graham@balena.io>

---

The `Steps` component has been updated to support the concept of having **ordered** steps. This mode of operation is designed to be used when the user must complete a sequence of steps _in order_ and be aware of which step they are currently on.

Ordered Steps

The existing `Steps` functionality is unchanged. The only minor change is making the step link/text colors more consistent.

##### Contributor checklist
- [x] I have rebuilt the README with `npm run build:docs`
- [x] I have regenerated screenshots for any affected components with `npm run generate-screenshots`
- [x] I have regenerated jest snapshots for any affected components with `npm run jest -- -u`

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---

### CODE REVIEW

**Comment by grahammcculloch** [comment]
Not sure why there is a failing `List` snapshot test - I haven't touched any code that affects that as far as I know. cc @sradevski

File: src/components/Steps/index.tsx [hunk 1]
```diff
@@ -99,9 +228,30 @@ const FramelessSteps = ({ children, ...otherProps }: any) => {
...
+		}: InternalStepsProps,
 		ref,
 	) => {
+		if (ordered) {
+			if (typeof activeStepIndex !== 'number') {
+				throw new Error(
+					'You must specify the activeStepIndex for ordered Steps',
+				);
+			} else {
+				const steps = React.Children.toArray(children);
```
Comment (sradevski) [inline src/components/Steps/index.tsx:249]: There is also `React.Children.count(children)` which is performant.

File: src/components/Steps/index.tsx [hunk 2]
```diff
@@ -1,72 +1,181 @@
...
 `;
 
-const HeaderContainer = styled(Flex)`
-	min-height: ${props => px(props.theme.fontSizes[5])};
+const StepTxt = styled(Txt)<{ active: boolean }>`
+	color: ${props => (props.active ? '' : props.theme.colors.tertiary.light)};
 `;
 
-export const Step = ({ status, children, onClick }: StepProps) => {
+const StepLink = styled(Link)`
```
Comment (sradevski) [inline src/components/Steps/index.tsx:83]: Why should this be overridden? IMO all links should look the same, and if we think the current one doesn't look as expected, we need to update the Link component itself.

File: src/components/Steps/index.tsx [hunk 3]
```diff
@@ -1,72 +1,181 @@
...
-			)}
+			<StepIcon
+				ordered={Boolean(ordered)}
+				status={status}
+				active={Boolean(active)}
+				index={index}
+			/>
 			{onClick ? (
-				<Link ml={status === 'none' ? 0 : 2} onClick={onClick}>
+				<StepLink color="tertiary.light" onClick={onClick}>
```
Comment (sradevski) [inline src/components/Steps/index.tsx:139]: Same reasoning here, I don't think we should be overriding the color. If it is like this in the designs, we need to chat with Alexandra and update the Link component instead.

File: src/components/Steps/index.tsx [hunk 4]
```diff
@@ -1,72 +1,181 @@
...
+const isStepClickable = (
+	ordered: boolean,
+	steps: any[],
+	index: number,
+	activeStepIndex: number,
+) => {
+	if (!ordered) {
+		return true;
+	}
+	const allPreviousStepsComplete = every(steps.slice(0, index), {
```
Comment (sradevski) [inline src/components/Steps/index.tsx:164]: We can use `findIndex` where status is not completed, and compare that with the component index instead, so we don't have to recreate an array twice (here and below) (maybe the lodash method, the native one doesn't have the best support yet).

**Comment by grahammcculloch** [comment]
@sradevski thanks for your detailed review. I've applied all the changes you suggested in [this commit](https://github.com/balena-io-modules/rendition/pull/1094/commits/a940c7d448613fe3a106e4bf93227c51afcfdc96). 

Also note my [earlier comment](https://github.com/balena-io-modules/rendition/pull/1094#issuecomment-618188750) about the failing `List` snapsot test...

**Comment by grahammcculloch** [comment]
@sradevski 
> I think you can remove node_modules, reinstall them and run the snapshot generation again, it's probably a dependency that got changed (since we don't have a lockfile) that caused the diff.

So I've tried the following:
```
git checkout master
rm -rf node_modules/
npm install
npm run jest -- src/components/List/spec.js
```

And I still get the error saying the snapshot has changed (same as in the failing CI step in this PR).

I've tried clearing the jest cache (`npm run jest -- --clearCache`) but that doesn't fix it.

Also on a clean master branch (as above) if I run `npm run build` I get the following error:
```
node_modules/htmlparser2/lib/WritableStream.d.ts:12:15 - error TS2749: 'StringDecoder' refers to a value, but is being used as a type here.

12     _decoder: StringDecoder;
````

FYI this error goes away if I update the `@types/node` package dependency to `^12.0.0`. But doesn't explain why its not working on my machine with a clean checkout of master.

I'm running node v12.16.1, npm v6.13.4 on Ubuntu (Windows WSL).

**Comment by sradevski** [comment]
@grahammcculloch can you run `npm run generate-snapshots` and push? As I said, since we don't lock on specific versions, some dependency might have changed and you get that. Once you update the snapshots it should work as expected.

**Comment by sradevski** [comment]
@balena-ci retest

**Comment by sradevski** [comment]
@grahammcculloch I am not sure why but one screenshot fails, which really makes me rethink the decision to not have a lockfile. Anyhow for now try running `npm run generate-screenshots` and see if that solves the issue.

**Pull request merged**